### PR TITLE
Home chain config syncer types

### PR DESCRIPTION
**Pull request by asoliman92**
## Motivation

Plugins need to refresh HomeChain config regularly to know which oracles support which chains and the latest F for each chain.

## Solution

Add required types (configs, interfaces)

CCIP [PR](https://github.com/smartcontractkit/ccip/pull/978)
Jira [Ticket](https://smartcontract-it.atlassian.net/browse/CCIP-2466)

### CODE REVIEW

File: pkg/types/ccipocr3/config.go [hunk 1]
```diff
@@ -12,12 +16,6 @@ type CommitPluginConfig struct {
...
 	DestChain ChainSelector `json:"destChain"`
 
-	// FChain defines the FChain value for each chain. FChain is used while forming consensus based on the observations.
-	FChain map[ChainSelector]int `json:"fChain"`
-
-	// ObserverInfo is a map of oracle IDs to ObserverInfo.
-	ObserverInfo map[commontypes.OracleID]ObserverInfo `json:"observerInfo"`
-
 	// PricedTokens is a list of tokens that we want to submit price updates for.
 	PricedTokens []types.Account `json:"pricedTokens"`
```
Comment (asoliman92) [inline pkg/types/ccipocr3/config.go:20]: I think this should be part of ObservationConsensus struct not the plugin config. It seems that it will be changing and we need to reach consensus on it between observers in DON.
CC @dimkouv @makramkd

File: pkg/types/ccipocr3/common_types.go [hunk 2]
```diff
@@ -85,3 +86,29 @@ func (b *BigInt) UnmarshalJSON(p []byte) error {
 func (b BigInt) IsEmpty() bool {
 	return b.Int == nil
 }
+
+type P2PID [32]byte
+
+func (b P2PID) MarshalJSON() ([]byte, error) {
+	return json.Marshal(hex.EncodeToString(b[:]))
```
Comment (makramkd) [inline pkg/types/ccipocr3/common_types.go:93]: P2P ID's should not be marshaled as hex, they have their own string encoding scheme please see libocr's PeerID type in ragep2p/types/types.go

Do we need to redefine the P2PID here? What's the win over using what we have in libocr?

File: pkg/types/ccipocr3/common_types.go [hunk 3]
```diff
@@ -85,3 +86,29 @@ func (b *BigInt) UnmarshalJSON(p []byte) error {
...
+	}
+	decoded, err := hex.DecodeString(s)
+	if err != nil {
+		return err
+	}
+	copy(b[:], decoded)
+	return nil
+}
+
+func GetP2pID(id int) P2PID {
```
Comment (makramkd) [inline pkg/types/ccipocr3/common_types.go:109]: What's this function for?

File: pkg/types/ccipocr3/config.go [hunk 4]
```diff
@@ -1,9 +1,13 @@
 package ccipocr3
 
+import "C"
```
Comment (makramkd) [inline pkg/types/ccipocr3/config.go:3]: Why is this import here? This is only imported for cgo files

File: pkg/types/ccipocr3/config.go [hunk 5]
```diff
@@ -78,3 +64,148 @@ type ObserverInfo struct {
...
+
+func (cm *P2PToSupportedChains) UnmarshalJSON(data []byte) error {
+	temp := make(map[string]SupportedChains)
+	if err := json.Unmarshal(data, &temp); err != nil {
+		return err
+	}
+	//if temp == nil {
+	//	*cm = nil
+	//	return nil
+	//}
```
Comment (makramkd) [inline pkg/types/ccipocr3/config.go:78]: Can delete this if not used

File: pkg/types/ccipocr3/config.go [hunk 6]
```diff
@@ -78,3 +64,148 @@ type ObserverInfo struct {
...
+	*cm = make(P2PToSupportedChains)
+	for k, v := range temp {
+		var key P2PID
+		decoded, err := hex.DecodeString(k)
```
Comment (makramkd) [inline pkg/types/ccipocr3/config.go:82]: See my comment above about p2p id encoding

File: pkg/types/ccipocr3/config.go [hunk 7]
```diff
@@ -78,3 +64,148 @@ type ObserverInfo struct {
...
+
+func (cm P2PToSupportedChains) MarshalJSON() ([]byte, error) {
+	temp := make(map[string]SupportedChains)
+	for k, v := range cm {
+		temp[hex.EncodeToString(k[:])] = v
+	}
+	return json.Marshal(temp)
+}
+
+type ConsensusObservation struct {
```
Comment (makramkd) [inline pkg/types/ccipocr3/config.go:100]: This name is too general, I suggest calling it something more specific since we come to consensus on a lot of things

File: pkg/types/ccipocr3/config.go [hunk 8]
```diff
@@ -78,3 +64,148 @@ type ObserverInfo struct {
...
+
+	return nil
+}
+
+type HomeChainConfig struct {
+	// FChain defines the FChain value for each chain. FChain is used while forming consensus based on the observations.
+	FChain map[ChainSelector]int `json:"fChain"`
+	// NodeSupportedChains is a map of PeerIDs to SupportedChains.
+	NodeSupportedChains P2PToSupportedChains `json:"nodeSupportedChains"`
+}
```
Comment (makramkd) [inline pkg/types/ccipocr3/config.go:129]: Why not embed this in the struct above?

File: pkg/types/ccipocr3/config.go [hunk 9]
```diff
@@ -78,3 +64,148 @@ type ObserverInfo struct {
...
+		}
+	}
+	if len(c.PricedTokens) == 0 {
+		return fmt.Errorf("priced tokens not set, at least one priced token is required")
+	}
+
+	return nil
+}
+
+type HomeChainConfig struct {
```
Comment (makramkd) [inline pkg/types/ccipocr3/config.go:124]: Lets add a comment to this and other structures to make it clear what they represent and what they're used for

File: pkg/types/ccipocr3/config.go [hunk 10]
```diff
@@ -78,3 +64,148 @@ type ObserverInfo struct {
...
+	supportedSlice := make([]ChainSelector, 0, sc.Supported.Cardinality())
+	for ch := range sc.Supported.Iter() {
+		supportedSlice = append(supportedSlice, ch)
+	}
+
+	// Define a temporary struct with a slice for Supported
+	temp := struct {
+		Supported []ChainSelector `json:"supported"`
+	}{
+		Supported: supportedSlice,
```
Comment (makramkd) [inline pkg/types/ccipocr3/config.go:196]: Is this deterministically generated in the same way all the time? If this is using a `map` under the hood iteration order is undefined so you might get different lists on different nodes and hence consensus failure.

If there is a risk of this we should `sort` the slice first before plugging it into this struct.

File: pkg/types/ccipocr3/config.go [hunk 11]
```diff
@@ -78,3 +64,148 @@ type ObserverInfo struct {
...
+
+	return json.Marshal(temp)
+}
+
+type OnChainConfig struct {
+	Readers []P2PID `json:"readers"`
+	FChain  uint8   `json:"fChain"`
+	Config  []byte  `json:"config"`
+}
+type OnChainCapabilityConfig struct {
```
Comment (makramkd) [inline pkg/types/ccipocr3/config.go:207]: This is not a capability configuration but a chain configuration

File: pkg/types/ccipocr3/config.go [hunk 12]
```diff
@@ -78,3 +64,148 @@ type ObserverInfo struct {
...
+	temp := struct {
+		Supported []ChainSelector `json:"supported"`
+	}{
+		Supported: supportedSlice,
+	}
+
+	return json.Marshal(temp)
+}
+
+type OnChainConfig struct {
```
Comment (makramkd) [inline pkg/types/ccipocr3/config.go:202]: Lets just have one struct called `ChainConfig` that has `Readers`, `FChain`, `Config`, `ChainSelector`, etc. The structure of offchain code does not have to mirror on-chain code

File: pkg/types/ccipocr3/interfaces.go [hunk 13]
```diff
@@ -53,3 +53,9 @@ type ExecutePluginCodec interface {
 type MessageHasher interface {
 	Hash(context.Context, CCIPMsg) (Bytes32, error)
 }
+
+type HomeChainPoller interface {
+	GetConfig() HomeChainConfig
```
Comment (makramkd) [inline pkg/types/ccipocr3/interfaces.go:58]: This needs to be either scoped to a particular chain selector or we get all the chain configs in one call. I think having both is more flexible at this stage.

```suggestion
	GetChainConfig(ctx context.Context, chainSelector uint64) HomeChainConfig
	GetAllChainConfigs(ctx context.Context) map[uint64]HomeChainConfig
```

File: pkg/types/ccipocr3/interfaces.go [hunk 14]
```diff
@@ -53,3 +53,9 @@ type ExecutePluginCodec interface {
...
+	Start(ctx context.Context) error
+	Close() error
```
Comment (makramkd) [inline pkg/types/ccipocr3/interfaces.go:60]: Probably better to inherit the `services.ServiceCtx` interface instead of doing this to make it clear what we want to implement

**Review by makramkd — CHANGES_REQUESTED** [review]

File: pkg/types/ccipocr3/interfaces.go (hunk 14 above)
Comment (asoliman92) [reply pkg/types/ccipocr3/interfaces.go:60]: I tried but it's a cyclic dependency unfortunately :/

File: pkg/types/ccipocr3/common_types.go (hunk 3 above)
Comment (asoliman92) [reply pkg/types/ccipocr3/common_types.go:109]: Might not be a good place. I'm using it in tests. Will remove and add to tesethelpers in ccip.

File: go.mod [hunk 15]
```diff
@@ -4,6 +4,7 @@ go 1.21
 
 require (
 	github.com/confluentinc/confluent-kafka-go/v2 v2.3.0
+	github.com/deckarep/golang-set/v2 v2.1.0
```
Comment (jmank88) [inline go.mod:7]: Does this library do anything sophisticated? Or could we use `map[T]struct{}` with `maps.Contains`/`maps.Keys`?

File: pkg/types/ccipocr3/common_types.go (hunk 2 above)
Comment (asoliman92) [reply pkg/types/ccipocr3/common_types.go:93]: I remember trying the libocr's PeerID first and then changing my mind for some reason. Let me check again why.

Comment (asoliman92) [reply pkg/types/ccipocr3/common_types.go:93]: I was using `PeerID` from `github.com/smartcontractkit/chainlink/v2/core/services/p2p/types` and this seems like a cyclic dependency.

Comment (makramkd) [reply pkg/types/ccipocr3/common_types.go:93]: Yeah don't import from core but you can import from libocr

Comment (makramkd) [reply pkg/types/ccipocr3/common_types.go:93]: Bonus is that libocr already implements the required Marshal* methods so we don't need that boilerplate again

File: pkg/types/ccipocr3/interfaces.go (hunk 14 above)
Comment (makramkd) [reply pkg/types/ccipocr3/interfaces.go:60]: Why is it cyclic? That is part of chainlink-common

File: pkg/types/ccipocr3/common_types.go [hunk 16]
```diff
@@ -85,3 +86,29 @@ func (b *BigInt) UnmarshalJSON(p []byte) error {
 func (b BigInt) IsEmpty() bool {
 	return b.Int == nil
 }
+
+type P2PID [32]byte
```
Comment (dimkouv) [inline pkg/types/ccipocr3/common_types.go:90]: There is also a `Bytes32` type

File: pkg/types/ccipocr3/config.go (hunk 8 above)
Comment (asoliman92) [reply pkg/types/ccipocr3/config.go:127]: I wanted them to be separate as they might grow differently in the future.

File: pkg/types/ccipocr3/config.go (hunk 12 above)
Comment (asoliman92) [reply pkg/types/ccipocr3/config.go:200]: How are we going to se/deserialize the actual on chain config that we get from the contract in this case?

File: pkg/types/ccipocr3/config.go (hunk 10 above)
Comment (asoliman92) [reply pkg/types/ccipocr3/config.go:194]: very good observation

File: go.mod (hunk 15 above)
Comment (asoliman92) [reply go.mod:7]: I'm using it for thread safe `Set` not for `map`

File: pkg/types/ccipocr3/config.go (hunk 12 above)
Comment (makramkd) [reply pkg/types/ccipocr3/config.go:200]: Deserialization of the contract data is done by the chain reader

File: pkg/types/ccipocr3/interfaces.go (hunk 14 above)
Comment (makramkd) [reply pkg/types/ccipocr3/interfaces.go:60]: You can just use `services.Service` I think, its in chainlink-common.

File: pkg/types/ccipocr3/config.go (hunk 10 above)
Comment (asoliman92) [reply pkg/types/ccipocr3/config.go:194]: We talked about this, the plugin is already using a Set as well, comparing sets is easy using the library and doesn't need to be in a specific order or anything.

File: pkg/types/ccipocr3/common_types.go (hunk 16 above)
Comment (asoliman92) [reply pkg/types/ccipocr3/common_types.go]: I removed this completely and using libocr PeerID 👍

**Review by makramkd — DISMISSED** [review]

**Review by makramkd — DISMISSED** [review]

File: pkg/types/ccipocr3/interfaces.go [hunk 17]
```diff
@@ -53,3 +58,15 @@ type ExecutePluginCodec interface {
 type MessageHasher interface {
 	Hash(context.Context, CCIPMsg) (Bytes32, error)
 }
+
+type HomeChainPoller interface {
+	GetChainConfig(chainSelector ChainSelector) (ChainConfig, error)
+	GetAllChainConfigs() (map[ChainSelector]ChainConfig, error)
+	// GetSupportedChainsForPeer Gets all chain selectors that the peerID can read/write from/to
+	GetSupportedChainsForPeer(id libocrtypes.PeerID) (mapset.Set[ChainSelector], error)
```
Comment (winder) [inline pkg/types/ccipocr3/interfaces.go:66]: How widely adopted is mapset? Generally I would avoid dependency types as part of a shared interface. Go uses `map[ChainSelector]bool` (or `map[ChainSelector]struct{}`) for sets.

**Review by 0xnogo — APPROVED** [review]

File: pkg/types/ccipocr3/interfaces.go (hunk 17 above)
Comment (jmank88) [reply pkg/types/ccipocr3/interfaces.go:66]: Right, this does not make sense to send over the wire. We could even return a plain `[]ChainSelector` and let the caller construct a set only if necessary.

**Review by jmank88 — CHANGES_REQUESTED** [review]

File: pkg/types/ccipocr3/interfaces.go [hunk 18]
```diff
@@ -53,3 +58,15 @@ type ExecutePluginCodec interface {
 type MessageHasher interface {
 	Hash(context.Context, CCIPMsg) (Bytes32, error)
 }
+
+type HomeChainPoller interface {
```
Comment (jmank88) [inline pkg/types/ccipocr3/interfaces.go:62]: Where is this used?

File: go.mod (hunk 15 above)
Comment (jmank88) [reply go.mod:7]: The idiomatic representation for a set in Go is a map with either `struct{}` or `bool` values.
You can see the implementation of the module imported here is doing the same thing: https://github.com/deckarep/golang-set/blob/main/threadunsafe.go#L34

There are interesting set operations in this library, but we don't seem to be using them, so the import may not be worth it. Although more importantly, it doesn't typically make sense to send types like this over grpc, since we want to pass data, not provide shared access to the raw data structures that we are holding on to.

**Comment by jmank88** [comment]
Does this work have a ticket and/or corresponding core/ccip PR?

File: pkg/types/ccipocr3/interfaces.go (hunk 18 above)
Comment (asoliman92) [reply pkg/types/ccipocr3/interfaces.go:62]: https://github.com/smartcontractkit/ccip/actions/runs/9614422103/job/26519178069?pr=978

Comment (asoliman92) [reply pkg/types/ccipocr3/interfaces.go:62]: It will be used in all CCIP plugins to get the dynamic configs from Home Chain

File: pkg/types/ccipocr3/interfaces.go (hunk 17 above)
Comment (asoliman92) [reply pkg/types/ccipocr3/interfaces.go:66]: good question, I saw it used in a couple of places in CCIP. Didn't notice that it was that little though.

File: pkg/types/ccipocr3/config.go [hunk 19]
```diff
@@ -78,3 +62,13 @@ type ObserverInfo struct {
 	// Reads define the chains that the current node can read from.
 	Reads []ChainSelector `json:"reads"`
 }
+
+// ChainConfig will live on the home chain and will be used to update chain configuration like F value and supported nodes dynamically.
+type ChainConfig struct {
```
Comment (jmank88) [inline pkg/types/ccipocr3/config.go:67]: Is this type also unused from this module?

Comment (asoliman92) [reply pkg/types/ccipocr3/config.go:67]: Yes. Will move it now 👍

**Review by jmank88 — APPROVED** [review]

**Review by 0xnogo — APPROVED** [review]
